src/Microsoft.AspNet.Hosting/HostingEngine.cs

```diff
@@ -56,14 +59,20 @@ public virtual IDisposable Start()
 
             var applicationDelegate = BuildApplicationDelegate();
 
+            var _logger = _applicationServices.GetRequiredService<ILogger<HostingEngine>>();
             var _contextFactory = _applicationServices.GetRequiredService<IHttpContextFactory>();
             var _contextAccessor = _applicationServices.GetRequiredService<IHttpContextAccessor>();
             var server = _serverFactory.Start(_serverInstance,
                 features =>
                 {
-                    var httpContext = _contextFactory.CreateHttpContext(features);
-                    _contextAccessor.HttpContext = httpContext;
-                    return applicationDelegate(httpContext);
+                    var requestIdentifier = GetRequestIdentifier(features);
```

Comment:
This needs to be re-arranged. At this point there's no grantee the feature collection is mutable so this can fail if you try to add to it (like on Kestrel?).  The context factory wraps this in a FeatureCollection to ensure it's mutable (https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNet.Hosting/Builder/HttpContextFactory.cs#L13). You should move `_contextFactory.CreateHttpContext(features);` up and then pass the HttpContext into GetRequestIdentifier.